idaes/apps/uncertainty_propagation/tests/test_uncertainties.py

```diff
@@ -0,0 +1,472 @@
+        # dc2/dp2 = -1
+        dc_dp[1,1] = -1
+
+        ### Uncertainty propagation
+        '''
+        Now lets test the uncertainty propagation package. We will assume p has covariance
+        sigma_p = [[2, 0], [0, 1]]
+        '''
+
+        ## Prepare inputs
+        # Covariance matrix
+        sigma_p = np.array([[2, 0], [0, 1]])
+
+        # Nominal values for uncertain parameters
+        theta = {'p1':m.p1(), 'p2':m.p2()}
+
+        # Names of uncertain parameters
+        theta_names = ['p1','p2']
+
+        # Important to unfix the parameters!
+        # Otherwise k_aug will complain about too few degrees of freedom
+        m.p1.unfix()
+        m.p2.unfix()
+
+        ## Run package
+        results = propagate_uncertainty(m, theta, sigma_p, theta_names)
+        
+        ## Check results
+
+        tmp_f = (df_dp + df_dx @ dx_dp)
+        sigma_f = tmp_f @ sigma_p @ tmp_f.transpose()
+
+        tmp_c = (dc_dp + dc_dx @ dx_dp)
+        sigma_c = tmp_c @ sigma_p @ tmp_c.transpose()
+
+        # This currently just checks if the order of the outputs did not change
+        # TODO: improve test robustness by using this information to set
+        # var_idx and theta_idx. This way the test will still work
+        # regardless of the order. In other words, the analytic solution needs to be
+        # reordered to match the variable/constraint order from
+        # this package. Alternately, the results could be converted into a Pandas dataframe
+        assert results.col == ['x1', 'x2', 'p1', 'p2', 'x3']
+        assert results.row == ['con1', 'con2', 'obj']
+        var_idx = np.array([True,True,False,False,True])
+        theta_idx = np.array([False,False,True,True,False])
+
+        # Check the gradient of the objective w.r.t. x matches
+        np.testing.assert_array_almost_equal(results.gradient_f[var_idx], np.array(df_dx))
+        
+        # Check the gradient of the objective w.r.t. p (parameters) matches
+        np.testing.assert_array_almost_equal(results.gradient_f[theta_idx], np.array(df_dp))
+        
+        # Check the Jacobian of the constraints w.r.t. x matches
+        np.testing.assert_array_almost_equal(results.gradient_c.toarray()[:, var_idx], np.array(dc_dx))
+        
+        # Check the Jacobian of the constraints w.r.t. p (parameters) matches
+        np.testing.assert_array_almost_equal(results.gradient_c.toarray()[:, theta_idx], np.array(dc_dp))
+        
+        # Check the NLP sensitivity results for the variables (x) matches
+        np.testing.assert_array_almost_equal(results.dsdp.toarray()[var_idx,:], np.array(dx_dp))
+        
+        # Check the NLP sensitivity results for the parameters (p) matches
+        np.testing.assert_array_almost_equal(results.dsdp.toarray()[theta_idx,:], np.array([[1,0],[0,1]]))
+        
+        # Check the uncertainty propagation results for the constrains matches
+        assert results.propagation_c == pytest.approx(np.sum(sigma_c))
```

Comment:
@JanghoPark-LBL `results.propagation_c` is a numpy array. Why not use `np.testing.assert_array_almost_equal` here?

Reply:
I updated .it